### ISSUE TITLE
Android build heavily pollutes source tree

### ISSUE THREAD

**Issue by mvglasow** [navit-gps/navit#762]
Currently, when we build Navit for Android using `scripts/build_android.sh` (which is now our common entry point for all builds—local, CI and F-Droid), build artifacts are created all over the source tree. Eariier versions of the script would even rename, move or modify some image files (which I have since fixed).

This makes cleaning up quite a nightmare. It involves deleting a list of directories and files, as well as resetting the Git repo. If you have uncommitted changes (a common occurrence when you are developing a new feature and are trying things out), there is a high risk of accidentally losing your changes.

Any way we can improve this? I would suggest aiming for the following:
* Do not mix source files (which includes images, makefiles etc.) and build artifacts in the same directory.
* Do not modify, rename, move or delete source files during build. If you must, create a copy and work on that (the copy would then be a build artifact)—currently OK but should be kept that way.
* When generating build artifacts inside the source tree, do so in as few locations as possible and only as many as strictly necessary. One or two would be ideal. (A location can be an entire directory hierarchy, as long as the root of the tree is safe to delete after build.)

**Comment by mvglasow** [comment]
Can anyone help me with that, or at least point me to the places where I would need to make these settings? It seems to be only Android that is affected (note that I do the Linux build from Eclipse, though). I’m not really a toolchain expert…

**Comment by mvglasow** [comment]
I just noticed `build_linux.sh` has the following lines:

```
BUILD_PATH="linux"

[ -d $BUILD_PATH ] || mkdir -p $BUILD_PATH
pushd $BUILD_PATH

# Build everything
    echo "Building..."
cmake ${cmake_opts} ../

```

…whereas `build_android.sh` runs cmake from the current dir. I’ll try adding the same there. Then I figure Gradle needs to be told to look for the native build artifacts in a different place. There are also some calls to `make` for icons, translations and XML which would need to be sorted out. No guarantee doing so won’t break anything—if someone has any advice for me (or, unlike me, knows what they’re doing), any pointers would be appreciated.

**Comment by mvglasow** [comment]
I notice the Android build runs cmake twice: once directly, close to the beginning of the script, and then again from gradle (close to the end of the script). Both instances generate a lot of files all over the source tree (as well as the other steps in between, i.e. generating images and XML files).

* What do the two `cmake` calls do/build? Do we really need both, or can we merge them into one?
* Can we cd to a build directory before invoking each? (How do I do that from inside gradle? Will doing so break the build?)

**Comment by mvglasow** [comment]
I've attempted to just cd into a separate build dir before invoking the first cmake, and change paths accordingly; results are in the `android-builddir` branch.

Currently, this fails at the first invocation of `gradlew` with

    Project directory '/home/michael/src/navit/android-builddir' is not part of the build defined by settings file '/home/michael/src/navit/settings.gradle'. If this is an unrelated build, it must have its own settings file.

What do I need to change so I can invoke `gradlew` from `android-builddir`?

**Comment by mvglasow** [comment]
As for the two cmake calls: @jkoan has confirmed my suspicion that the first call is needed to generate the makefiles we need to build translations, XML config files and bitmap images in the next step, which is done with `make`. These things need to be in place before Gradle assembles the APK. However, this currently comes at the cost of building the entire set of binaries before running gradle, and never using them. @jkoan tells me he’s currently working on getting rid of the double CMake.

As for the build dir: According to the [docs](https://developer.android.com/reference/tools/gradle-api/4.1/com/android/build/api/dsl/Cmake#buildstagingdirectory), `cmake` lets us specify a `buildStagingDirectory`, i.e. where CMake will store its output. Could be useful for our purposes, though I haven’t figured out what erxactly this does: tell CMake to drop its generated files there, or just tell Gradle to pick the binaries up there?

If this controls both, a solution might be to cd out of the build dir before invoking gradle, and setting `buildStagingDirectory` to point to the build dir.

**Comment by mvglasow** [comment]
Partial success, at least as far as CMake output goes. Running gradlew from the source dir works, and with `buildStagingDirectory` pointing to the build dir (which is directly underneath the root of the source tree), there are no more CMake artifacts in the source tree, except for one single file. The generated APK looks OK—Navit starts up, and all bitmap files seem to be on board.

Now tbe build leftovers are down to five items:
* `android-builddir/` (the build dir)
* `cmake_plugin_settings.txt`
* `navit/android/assets/` (XML config files, entire dir)
* `navit/android/res/drawable-nodpi/` (PNG bitmaps, entire dir)
* `navit/android/res/raw/`(translations, entire dir)

**Comment by mvglasow** [comment]
As for the leftovers in navit/android, these are copied over from the build dir. Paths map as follows:

Build dir | Source tree | Other files
-|-|-
`./android-builddir/navit/icons/` | `./navit/android/res/drawable-nodpi/` | CMake files, desktop_icons, *.svg
`./android-builddir/navit/config` | `./navit/android/assets/config` | None
`./android-builddir/po` | `./navit/android/res/raw` | CMake files, *.po, navit_shipped.c, string_xml.c, navit.pot

We could try two things to keep all build artifacts inside the build dir:
* Rather than copying stuff to the source tree, designate a location in the build dir for each type of file and copy the files there. In `build.gradle`, extend the set of dirs to include these. Potential obstacle: we’d need to have multiple resource dirs (one for those tracked by git, one for generated ones), not sure if this is possible
* Do not copy the files at all, but point `build.gradle` to wherever make has dropped these files.
  * Potential benefit: If Gradle processes the makefiles in some way, it would also build the xml/png/mo files, allowing us to ditch the CMake/make jobs outside of Gradle.
  * Potential obstacles: The resource files we need are mixed with other stuff in the build dirs, which might confuse the Android toolchain. Also, `icons` and `po` do not conform to the directory structure which Android expects—not sure if ad how we can work around those two.

**Comment by mvglasow** [comment]
Another obstacle for option 2: we are currently renaming some of these files (bitmaps and translations), converting any uppercase to lowercase letters. That would need to be solved somehow as well…

**Comment by mvglasow** [comment]
#1094 now includes option 1. Apart from one single settings file (which we could add to `.gitignore`), all build artifacts stay inside the build dir.

Pursuing option 2 would be beneficial only if we want to eliminate the extra CMake/make before running gradle.

**Comment by mvglasow** [comment]
Since #1094 (with the second commit) fully resolves this issue, I am closing it.
